Feature/50266 conditionally remove shares of removed project members

toy opened the pull request:
Script to generate possible members (first run `bin/rails backup:database:create'[dump/initial]'` to create snapsot, then `bin/rails runner PATH_TO_FILE` to generate):
```ruby
Kernel.system(OpenProject::PostgresEnvironment.pg_env, *%W[
  pg_restore
  --clean
  --no-owner
  --single-transaction
  --dbname=#{OpenProject::PostgresEnvironment.pg_env['PGDATABASE']}
  dump/initial
]) if File.exist?('dump/initial')

include FactoryBot::Syntax::Methods

admin = User.find_by(login: 'admin')

project = create(:project)
project_role_direct = create(:project_role, name: 'direct')
project_role_inherited = create(:project_role, name: 'inherited')
view_work_package_role = WorkPackageRole.find_by(name: 'Work package viewer') || create(:view_work_package_role)
comment_work_package_role = WorkPackageRole.find_by(name: 'Work package commenter') || create(:comment_work_package_role)
edit_work_package_role = WorkPackageRole.find_by(name: 'Work package editor') || create(:edit_work_package_role)

name_sequence = ('AA'..).to_enum

[true, false].repeated_permutation(4) do |direct_project_role, inherited_project_role, direct_shared_wps, inherited_shared_wps|
  next unless direct_project_role || inherited_project_role || direct_shared_wps || inherited_shared_wps

  name = name_sequence.next

  user = create(:user, firstname: "#{name}", lastname: [
    direct_project_role ? 'DirRole' : nil,
    inherited_project_role ? 'GroupRole' : nil,
    direct_shared_wps ? 'DirectShared' : nil,
    inherited_shared_wps ? 'GroupShared' : nil,
  ].compact.join(' '))

  if inherited_project_role || inherited_shared_wps
    group = create(:group, lastname: "G_#{name}", members: [user])
  end

  if direct_project_role
    create(:member, principal: user, project:, roles: [project_role_direct])
  end

  if inherited_project_role
    create(:member, principal: group, project:, roles: [project_role_inherited])
  end

  if direct_shared_wps || inherited_shared_wps
    work_package_both = create(:work_package, project:)
  end

  if direct_shared_wps
    work_package_direct = create(:work_package, project:)

    create(:member, principal: user, project:, roles: [view_work_package_role], entity: work_package_direct)
    create(:member, principal: user, project:, roles: [edit_work_package_role], entity: work_package_both)
  end

  if inherited_shared_wps
    work_package_inherited = create(:work_package, project:)

    create(:member, principal: group, project:, roles: [comment_work_package_role], entity: work_package_inherited)
    create(:member, principal: group, project:, roles: [edit_work_package_role], entity: work_package_both)
  end

  if inherited_project_role || inherited_shared_wps
    Groups::CreateInheritedRolesService
      .new(group, current_user: create(:admin))
      .call(user_ids: group.user_ids,
            send_notifications: false,
            project_ids: [project.id])
  end
end
```

------

https://community.openproject.org/wp/50266

aaron-contreras commented on app/components/members/row_component.rb:
```diff
@@ -152,6 +152,11 @@ def edit_link
       )
     end
 
+    def edit_icon
+      render(Primer::Beta::Octicon.new(icon: :pencil))
+      # render(Primer::Beta::IconButton.new(size: :small, icon: :pencil, "aria-label": "Edit roles"))
+    end
+
```
Compliment on your "table of contents" ordering of methods

toy committed:
restore marking disabled inherited roles in member role editing

This was accidentally removed when convert cells to components in 1d541f82373191238af474633560bef80abd7b22

toy committed:
allow adding roles for members that have only shared work packages

aaron-contreras commented on app/components/members/delete_dialog_component.html.erb:
```diff
@@ -40,7 +40,7 @@ See COPYRIGHT and LICENSE files for more details.
       <%= render(Primer::Beta::Button.new(data: { close_dialog_id: id })) { "Cancel" } %>
       <%= render(Primer::Beta::Button.new(scheme: :danger, tag: :a, href: delete_url(delete_member: true), data: { method: :delete })) { "Keep only shared" } %>
-      <%= render(Primer::Beta::Button.new(scheme: :danger, tag: :a, href: delete_url(delete_member: true, delete_shared_work_packages: true), data: { method: :delete })) { "Remove all" } %>
+      <%= render(Primer::Beta::Button.new(scheme: :danger, tag: :a, href: delete_url(delete_member: true, delete_work_package_shares: true), data: { method: :delete })) { "Remove all" } %>
```
Thanks, this is no longer misleading.

toy replied:
Noted while fixing specs that this reads wrong

toy committed:
correct rubocop fix of should in example descriptions

toy committed:
prefer quoted_table_name over just table_name

toy committed:
modify test to show that double membership gives wrong count when filtering by role_id

toy committed:
fix wrong count when filtering double membership by role_id

toy committed:
ensure 0 is returned instead of NULL for members without shared work packages

toy committed:
remove unused let variable other_project

toy committed:
extend with_shared_work_packages_count to provide also inherited and totals

toy committed:
no need to have explicit shared_work_packages_count method

toy committed:
delete non inherited roles for members with inherited roles in members delete service

toy committed:
remove unused let global in spec/helpers/work_packages_filter_helper_spec.rb

toy committed:
add test for WorkPackagesFilterHelper#project_work_packages_shared_with_path

toy committed:
add WorkPackagesFilterHelper#project_work_packages_with_ids_path as is currently more difficult to add filtering by role

toy committed:
get shared_work_package_ids instead of just count

toy committed:
default to showing work packages shared with principal, switch to ids when required

toy committed:
don't show shared work packages if current user is not allowed to view them

toy committed:
add Member#some_roles_deletable? instead of changing deletable?

toy committed:
rework interface for managing members

toy committed:
rubocop

toy committed:
a bit of cleanup in Pages::Members

ulferts commented on app/components/members/delete_member_dialog_component.html.erb:
```diff
@@ -0,0 +1,85 @@
+      <%= render(Primer::Beta::Button.new(scheme: :danger, tag: :a, href: delete_url(project: true, work_package_shares_role_id: "all"), data: { method: :delete })) { t(:button_remove_member_and_shares) } %>
+    <% end %>
+  <% else %>
+    <% if can_delete_roles? %>
```
This can be turned into a single `elsif` to avoid one layer of else-if

ulferts commented on app/components/members/delete_member_dialog_component.html.erb:
```diff
@@ -0,0 +1,85 @@
+    <% else %>
+      <% dialog.with_body do %>
+        <%= render(Primer::Beta::Text.new(tag: "p")) { scoped_t(:cannot_delete_group_member) } %>
+        <%= render(Primer::Beta::Text.new(tag: "p")) { scoped_t(:cannot_delete_group_member_note_html, administration_settings_link:) } %>
```
In case the user is no admin, referencing the admin group settings will lead to a link those users are not allowed to follow. I'd rather only show that link if the user is an admin and can actually do something about the group membership.

ulferts commented on app/components/members/delete_member_dialog_component.html.erb:
```diff
@@ -0,0 +1,85 @@
+  <% if can_delete_roles? && can_delete_shares? %>
+    <% if can_delete? %>
+      <% dialog.with_body do %>
+        <%= render(Primer::Beta::Text.new(tag: "p")) { scoped_t(:will_remove_the_users_role_html, shared_work_packages_link:) } %>
```
I like the usage of `scoped_t` here. Definitely improves readability.

ulferts replied:
The `if` statement can be moved inside the `dialog.with_body` as both cases will just change the content of the body. That would make it a bit easier to understand.

ulferts commented on app/components/members/delete_work_package_shares_dialog_component.html.erb:
```diff
@@ -0,0 +1,112 @@
+    <% if direct_shared_work_packages_count? %>
+      <% if inherited_shared_work_packages_count? %>
+        <% dialog.with_body do %>
+          <%= render(Primer::Beta::Text.new(tag: "p")) do %>
```
You could extract the rather repetitive `render(Primer::Beta::Text.new(tag: "p"))` into a method like `render_paragraph` or `render_text` or `primer_text`

toy replied:
I was thinking about doing that

ulferts commented on app/components/members/delete_work_package_shares_dialog_component.html.erb:
```diff
@@ -0,0 +1,112 @@
+      <% end %>
+    <% end %>
+  <% else %>
+    <% if direct_shared_work_packages_count? %>
```
Here, you could again avoid one nesting when using `elsif` instead.

ulferts commented on app/components/members/delete_work_package_shares_dialog_component.html.erb:
```diff
@@ -0,0 +1,112 @@
+      <% end %>
+    <% end %>
+  <% end %>
+<% end %>
```
These deeply nested conditions are hard to follow and I don't have a recipe I am sure will work to improve the situation. But what I would try is avoiding nested `if` in favor of `elsif` wherever the full content of the modal (body + footer) differentiates between the cases. That would require repeating some of the conditions but I believe that this is actually beneficial for readability. Next thing I am rather unsure about but might be worth a try is trying to label the different variants with something like `only_direct_shares?` and `direct_and_inherited_shares?`. With that in place I would revisit whether it makes sense to deliberately accept repetitions in favor of having the different variants stand out more clearly (and being able to introduce more states) by having both the body as well as the footer completely mentioned within each state. To avoid repetition, it might then be necessary to extract repeating parts into the methods.

But I really don't know, if that is a good change and would have to try it out in order to access whether it improves he situation.

toy replied:
I've reduced nesting (but didn't remove it completely) and it looks cleaner, so have a look

ulferts commented on config/locales/en.yml:
```diff
@@ -345,6 +344,29 @@ en:
+      title: "Revoke work package shares"
+      shared_with_this_user_html: "%{all_shared_work_packages_link} have been shared with this user."
+      shared_with_this_group_html: "%{all_shared_work_packages_link} have been shared with this group."
+      shared_with_permission_html: "Only %{shared_work_packages_link} have been shared with %{shared_role_name} permissions."
```
These always have the plural "have" in them. In case only "1 work package" is shared, that reads oddly.

ulferts commented on config/locales/en.yml:
```diff
@@ -333,7 +333,6 @@ en:
       no_results_content_text: Add a member to the project
     invite_by_mail: "Send invite to %{mail}"
     send_invite_to: "Send invite to"
-    no_modify_on_shared: "You currently cannot modify or remove shared memberships through the member page. Use the sharing modal instead."
```
Bonus points for removing this key

ulferts commented on app/services/members/delete_service.rb:
```diff
@@ -29,6 +29,14 @@
+      object.member_roles.where(inherited_from: nil).destroy_all
+    end
+  end
+
```
In case the member is in fact not completely deleted, I would not send out a notification (`MEMBER_DESTROYED`). The member is still there after all.

toy replied:
Tell me if you have a better idea how to handle it then ff49068bde55501a9644ee6c94577c9097d603e3

ulferts commented on spec/services/members/delete_by_principal_service_spec.rb:
```diff
@@ -0,0 +1,214 @@
+               entity: work_package_c)
+      end
+      let!(:member_d) do
+        create(:member,
```
Are these `create` calls necessary or would a `build`/`build_stubbed` suffice?

toy replied:
In this case yes, as the service uses scopes to find what to delete and it should be covered in tests

ulferts reviewed (changes requested):
@toy, I overall like the PR. The structure looks good and the code is well written. 

As was to be expected when tackling a complex topic, there are a couple of issues in the PR: 

* If the user has the permission "Share work packages" but lacks the permission to "Manage members", the actions in the members list will still show the "delete action". Clicking on it will lead to an error.

* There is an edge case when: 
  * User A has a WP shared with "Comment" role
  * Group G has same WP shared with "View" role
  * Going to the "Comment" filter in the members list and choosing to revoke shares
The user can click on "Revoke all" but it will be the same as removing only the "Comment" shares. The ones of the roles are not removed. This might be an edge case but it feels weird:
<img width="663" alt="image" src="https://github.com/opf/openproject/assets/617519/59eab252-6396-4986-b591-cf4c648841a5">
We would need to discuss if we want to do something about it.

* No feature specs have been added. There are probably too many cases to cover in feature specs, but covering the main ones should be done. I did not check how well that part of the application was already covered by feature specs so take that statement with a grain of salt. 

* When removing a group, the modal says: "Deleting this member will remove all access privileges of the user to the project. The user will still exist as part of the instance." - speaking of a user but it is a group. No mentioning of how it will affect the group's members, the users. Ideally some text would explain that their access privileges are removed as well.

* A topic of personal preference probably, but I would prefer a bit more space between the header and the body of the dialog. In total, the dialogs as designed look a bit more lofty:
<img width="665" alt="image" src="https://github.com/opf/openproject/assets/617519/b6a72396-8d16-47ea-9ecc-f044c9307593">

compared to ones implemented:
<img width="637" alt="image" src="https://github.com/opf/openproject/assets/617519/591fcd63-fc33-4883-9e46-0bf892639f8e">

Which is due to the space between body and header but also between paragraphs in the body. I see in the implementation that the defaults were chosen so I don't blame this PR for it. But I guess that defining the layout of such confirmation dialogs a bit more would be fruitful /cc @HDinger 

* When removing project memberships not via the project members table but rather via the API, no confirmation dialog at all is displayed and only the project memberships are removed whereas the shares remain. Ideally, we would reuse the modals and have the user decide whether to keep/delete the shares as well.  

* The flash messages always report that the "member" was removed from the project. This is also true if only shares have been removed or when the member wasn't completely removed. This is a bit surprising in some scenarios. I would not want to change the way we display the messages for now but the content could be improved.

toy committed:
use the non existent destroy action for checking if user may delete member

Both deleting member and work package share require access to destroy_by_principal
So either we can use deleted destroy action or check not action access, this looks simpler

toy commented:
> If the user has the permission "Share work packages" but lacks the permission to "Manage members", the actions in the members list will still show the "delete action". Clicking on it will lead to an error.

Good catch, I first switched to checking access to `destroy_by_principal` action, but as it is needed also for deleting work package shares, it is not valid to use for permission to delete member.
See if you agree with https://github.com/opf/openproject/pull/14721/commits/80af7b8eeb43b50a227d6a157e204d6801d81cae

toy commented:
> * There is an edge case when:
>   
>   * User A has a WP shared with "Comment" role
>   * Group G has same WP shared with "View" role
>   * Going to the "Comment" filter in the members list and choosing to revoke shares
>     The user can click on "Revoke all" but it will be the same as removing only the "Comment" shares. The ones of the roles are not removed. This might be an edge case but it feels weird:
> 
> <img alt="image" width="663" src="https://private-user-images.githubusercontent.com/617519/316093405-59eab252-6396-4986-b591-cf4c648841a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMjk5MDcsIm5iZiI6MTcxMTEyOTYwNywicGF0aCI6Ii82MTc1MTkvMzE2MDkzNDA1LTU5ZWFiMjUyLTYzOTYtNDk4Ni1iNTkxLWNmNGM2NDg4NDFhNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyMlQxNzQ2NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMDBiNjM5OTdlYzIzNzE1NjY4OTgyMzZkZDcwZGFkYzcwOWQxMjdkMmM4N2YwNGMzODllOGZjNDY4ZTE0Y2EyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.jIU_yZ6P3d7O0aLFW3zA9t12UWNUZxpDO2WI1g8DUHM"> We would need to discuss if we want to do something about it.

I already discussed it with @marcalcobe and he suggested to treat it as an edge case.

toy commented:
> * No feature specs have been added. There are probably too many cases to cover in feature specs, but covering the main ones should be done. I did not check how well that part of the application was already covered by feature specs so take that statement with a grain of salt.

There were already some feature tests for deleting member that go through action menu and dialog, there are integration tests for controller that check all main parameters and services are completely covered. But tell me if you think that there should be feature tests for revoking work packages and/or revoking work packages alongside member.

toy commented:
> * When removing a group, the modal says: "Deleting this member will remove all access privileges of the user to the project. The user will still exist as part of the instance." - speaking of a user but it is a group. No mentioning of how it will affect the group's members, the users. Ideally some text would explain that their access privileges are removed as well.

Good catch, I handled the user vs group in revoke dialog, but not here. I will ask Marc for texts

toy commented:
> * A topic of personal preference probably, but I would prefer a bit more space between the header and the body of the dialog. In total, the dialogs as designed look a bit more lofty:
> 
> <img alt="image" width="665" src="https://private-user-images.githubusercontent.com/617519/316080441-b6a72396-8d16-47ea-9ecc-f044c9307593.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMjk3NDYsIm5iZiI6MTcxMTEyOTQ0NiwicGF0aCI6Ii82MTc1MTkvMzE2MDgwNDQxLWI2YTcyMzk2LThkMTYtNDdlYS05ZWNjLWYwNDRjOTMwNzU5My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyMlQxNzQ0MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MGJiNjc3MjhjZmY3NmRiNzcxMGNkYmU5NjYyNjc3NDA0ODU2NjQ4Y2Y4ZjQyNTMyY2Y0MTY2N2VkZjAwNmQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7Ow5OeZcBgOrIoKRVzUJ0PSLUXg0E5_SQXLIzBfmVr8">
> compared to ones implemented: <img alt="image" width="637" src="https://private-user-images.githubusercontent.com/617519/316080641-591fcd63-fc33-4883-9e46-0bf892639f8e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMjk3NDYsIm5iZiI6MTcxMTEyOTQ0NiwicGF0aCI6Ii82MTc1MTkvMzE2MDgwNjQxLTU5MWZjZDYzLWZjMzMtNDg4My05ZTQ2LTBiZjg5MjYzOWY4ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyMlQxNzQ0MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZmVlYzZiMmU0M2MxYjY4NGJkYzA2Yzk4N2RmMGRiNzM5MTcwNTg0ZGJmMjE3OWQwM2Q4YTQzM2IwNmY5YTljJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7WXhO4FN9V789dwpxyn3hG9xD0oVmMcGJ2AGqmekkVM">
> 
> Which is due to the space between body and header but also between paragraphs in the body. I see in the implementation that the defaults were chosen so I don't blame this PR for it. But I guess that defining the layout of such confirmation dialogs a bit more would be fruitful /cc @HDinger

I'm trying to follow using components as is without additional styling. Maybe @marcalcobe also can chime in

toy commented:
> * When removing project memberships not via the project members table but rather via the API, no confirmation dialog at all is displayed and only the project memberships are removed whereas the shares remain. Ideally, we would reuse the modals and have the user decide whether to keep/delete the shares as well.

I didn't get what you mean by API. Wouldn't API mean not Web UI?

toy commented:
> * The flash messages always report that the "member" was removed from the project. This is also true if only shares have been removed or when the member wasn't completely removed. This is a bit surprising in some scenarios. I would not want to change the way we display the messages for now but the content could be improved.

The texts of notifications was the thing I didn't like, but I though we agreed to leave it as is for now. @marcalcobe

HDinger commented on app/components/members/delete_member_dialog_component.html.erb:
```diff
@@ -0,0 +1,85 @@
+
+    <% dialog.with_footer do %>
+      <%= render(Primer::Beta::Button.new(data: { close_dialog_id: id })) { t(:button_cancel) } %>
+      <%= render(Primer::Beta::Button.new(scheme: :danger, tag: :a, href: delete_url(project: true), data: { method: :delete })) { t(:button_remove_member) } %>
+      <%= render(Primer::Beta::Button.new(scheme: :danger, tag: :a, href: delete_url(project: true, work_package_shares_role_id: "all"), data: { method: :delete })) { t(:button_remove_member_and_shares) } %>
```
<img width="400" alt="image" src="https://github.com/opf/openproject/assets/617519/591fcd63-fc33-4883-9e46-0bf892639f8e">

Primer recommends to use the `danger` buttons sparingly. So I am wondering whether we can avoid the doubled buttons here. Maybe with a checkbox before stating something like "remove shares as well". Opinions on that @marcalcobe @psatyal ?

toy replied:
We discussed using checkboxes, and even combining editing roles and deleting into one dialog with lots of checkboxes, but went back to buttons

marcalcobe replied:
We discussed this again in the Frontend daily this morning with all devs and agreed that we would need to modify this design again to try to follow as much as we can Primer. Together with @psatyal we created the mockups in this page: https://www.figma.com/file/lLDYzQW7evoBA4lha5PwXO/Project-members?type=design&node-id=1135-29390&mode=design

I know this is a change again on how the modal looks like and it should had been taken in consideration earlier, I'm sorry for that and I hope this doesn't cost to much extra time to implement.

marcalcobe commented:
> > * When removing a group, the modal says: "Deleting this member will remove all access privileges of the user to the project. The user will still exist as part of the instance." - speaking of a user but it is a group. No mentioning of how it will affect the group's members, the users. Ideally some text would explain that their access privileges are removed as well.
> 
> Good catch, I handled the user vs group in revoke dialog, but not here. I will ask Marc for texts

I've added some mockups for this case when trying to remove a group: https://www.figma.com/file/lLDYzQW7evoBA4lha5PwXO/Project-members?type=design&node-id=1136-13208&mode=design&t=Nxyc5ejAZ3jBQXUf-4

marcalcobe commented:
> > * A topic of personal preference probably, but I would prefer a bit more space between the header and the body of the dialog. In total, the dialogs as designed look a bit more lofty:
> > 
> > <img alt="image" width="665" src="https://private-user-images.githubusercontent.com/617519/316080441-b6a72396-8d16-47ea-9ecc-f044c9307593.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMjk3NDYsIm5iZiI6MTcxMTEyOTQ0NiwicGF0aCI6Ii82MTc1MTkvMzE2MDgwNDQxLWI2YTcyMzk2LThkMTYtNDdlYS05ZWNjLWYwNDRjOTMwNzU5My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyMlQxNzQ0MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MGJiNjc3MjhjZmY3NmRiNzcxMGNkYmU5NjYyNjc3NDA0ODU2NjQ4Y2Y4ZjQyNTMyY2Y0MTY2N2VkZjAwNmQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7Ow5OeZcBgOrIoKRVzUJ0PSLUXg0E5_SQXLIzBfmVr8">
> > compared to ones implemented: <img alt="image" width="637" src="https://private-user-images.githubusercontent.com/617519/316080641-591fcd63-fc33-4883-9e46-0bf892639f8e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTExMjk3NDYsIm5iZiI6MTcxMTEyOTQ0NiwicGF0aCI6Ii82MTc1MTkvMzE2MDgwNjQxLTU5MWZjZDYzLWZjMzMtNDg4My05ZTQ2LTBiZjg5MjYzOWY4ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyMlQxNzQ0MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZmVlYzZiMmU0M2MxYjY4NGJkYzA2Yzk4N2RmMGRiNzM5MTcwNTg0ZGJmMjE3OWQwM2Q4YTQzM2IwNmY5YTljJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.7WXhO4FN9V789dwpxyn3hG9xD0oVmMcGJ2AGqmekkVM">
> > Which is due to the space between body and header but also between paragraphs in the body. I see in the implementation that the defaults were chosen so I don't blame this PR for it. But I guess that defining the layout of such confirmation dialogs a bit more would be fruitful /cc @HDinger
> 
> I'm trying to follow using components as is without additional styling. Maybe @marcalcobe also can chime in

About this I would suggest we use the component as it comes from Primer to not complicate the implementation.

toy committed:
add missing route keys for delete urls in dialogs

toy committed:
add a test for DeleteMemberDialogComponent

toy committed:
add a test for DeleteWorkPackageSharesDialogComponent

toy committed:
use helpers for paragraph and button in both dialogs

toy committed:
reduce nesting in dialog htmls

toy committed:
use helper for cancel_button in both dialogs

toy committed:
separate will_remove_the_users_role_html into will_remove_the_users_role and however_shares_with_user_html

toy committed:
rename total_shared_work_packages_count to all_shared_work_packages_count

toy committed:
handle pluralisation for the rest of the message with link to work packages

toy committed:
show suggestion to use administration interface only when current user has permission for that

toy committed:
combine two cases of not being able to delete anything in DeleteWorkPackageSharesDialogComponent

toy committed:
separate texts in dialog when removing group member

toy committed:
don't send notification when only some project roles, but not member itself is deleted

toy commented on app/services/members/delete_service.rb:
```diff
@@ -44,7 +46,7 @@ def after_perform(service_call)
 
       cleanup_for_group(member)
-      send_notification(member)
+      send_notification(member) if @send_notification
```
use `.destroyed?`

toy committed:
renamed some dialog i18n keys for consistency and clarity

ulferts commented on app/components/members/delete_member_dialog_component.html.erb:
```diff
@@ -0,0 +1,97 @@
+      <%= cancel_button %>
+    <% end %>
+  <% end %>
+<% end %>
```
I like it @toy. It still requires heavy thinking but the combination of the case statement with the reduction due to `paragraph` at al. makes reading it simpler. 👏

toy committed:
don't use confusing can_delete_shares?

Use separate may_delete_shares? and shared_work_packages?
It was still true when all work package shares are inherited, so can't be deleted

ulferts commented:
> > * When removing project memberships not via the project members table but rather via the API, no confirmation dialog at all is displayed and only the project memberships are removed whereas the shares remain. Ideally, we would reuse the modals and have the user decide whether to keep/delete the shares as well.
> 
> I didn't get what you mean by API. Wouldn't API mean not Web UI?

For whatever reason, I definitely used the wrong term. What I meant to say is that there is no confirmation dialog when removing the project memberships from the administration (Administration -> Users and permissions -> "A User" -> Project tab):

<img width="837" alt="image" src="https://github.com/opf/openproject/assets/617519/ef6eb07d-67eb-4054-9389-5e1e7bc9b575">

toy commented:
> > > * When removing project memberships not via the project members table but rather via the API, no confirmation dialog at all is displayed and only the project memberships are removed whereas the shares remain. Ideally, we would reuse the modals and have the user decide whether to keep/delete the shares as well.
> > 
> > 
> > I didn't get what you mean by API. Wouldn't API mean not Web UI?
> 
> For whatever reason, I definitely used the wrong term. What I meant to say is that there is no confirmation dialog when removing the project memberships from the administration (Administration -> Users and permissions -> "A User" -> Project tab):
> 
> <img alt="image" width="837" src="https://private-user-images.githubusercontent.com/617519/317761825-ef6eb07d-67eb-4054-9389-5e1e7bc9b575.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE2Mzg1MDAsIm5iZiI6MTcxMTYzODIwMCwicGF0aCI6Ii82MTc1MTkvMzE3NzYxODI1LWVmNmViMDdkLTY3ZWItNDA1NC05Mzg5LTVlMWU3YmM5YjU3NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyOFQxNTAzMjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01N2I5ZTY4ZDNmOGViYmMwNDFhZmNhNzM4Yzg2ODg5NmQ3MjA3MzRkM2JmNGQ2OGQxNDc2NzkzNzJkYTNmNDcxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.gLgPMxJW1Vp9ZTdiJXlnb39Zo_FDJkJjgeC-9i0UMYU">

Ah yes, I agree that it should show the dialog, though I suggest doing it separately

ulferts reviewed (changes requested):
Hey @toy ,

thanks for addressing the issues I raised. Looks almost good now. Almost, since when retesting I found a weird behaviour which I think is a regression:

Every member row, regardless of it having a work package shared or not and regardless of it being a user (and not a group) has the option now to "Revoke work package shares", when clicking on it the following is displayed:

<img width="1470" alt="image" src="https://github.com/opf/openproject/assets/617519/e67679bf-a287-43f4-be22-e7d20a43cfaa">

Again, this was shown after clicking on the action item for a user.

toy commented:
> Every member row, regardless of it having a work package shared or not and regardless of it being a user (and not a group) has the option now to "Revoke work package shares", when clicking on it the following is displayed:

Good catch, indeed I broke it in the last commit

ulferts reviewed (approved):
🎉 Thanks for pushing through.
